### ISSUE TITLE
isEqualToComparingFieldByFieldRecursively does not check the class

### ISSUE THREAD

**Issue by GaspardPO** [joel-costigliola/assertj-core#1054]
#### Summary
when using isEqualToComparingFieldByFieldRecursively or isEqualToComparingFieldByField, the comparison is not done on class type, only on field value (and name).

So two POJO with the same fields, but from different classes are considered as equals.
It could be useful to test the object instance (recursively, or not)

#### Example

```java

    @Test
    public void test() {
        A a = new A(10);
        B b = new B(10);
        assertThat(a).isEqualToComparingFieldByFieldRecursively(b); 
 // I would like this line to fail because A and B are different classes
 // but it is not failing
    }

    private static class A {
        int x;

        public A(int x) {
            this.x = x;
        }
    }

    private static class B {
        int x;

        public B(int x) {
            this.x = x;
        }

    }

```

#### Java 8 specific ?
* NO  : create Pull Request from the `2.x` branch

**Comment by joel-costigliola** [comment]
It is by design, changing it would be a breaking change, you can use `isInstanceOf` to check the type.

We are planning to provide a new API for recursive comparison, can you add a comment on https://github.com/joel-costigliola/assertj-core/issues/1002 so that we don't forget to support your request ?

**Comment by GaspardPO** [comment]
I added a comment in #1002 

And BTW, a lot of thanks for all the work, it is a really well done API, and very useful.

**Comment by joel-costigliola** [comment]
Thanks for the kind words, I'm closing this issue as it will be addressed in #1002.
